pyccel/ast/core.py

```diff
@@ -1252,7 +1253,103 @@ def is_symbolic_alias(self):
 
         return False
 
+#------------------------------------------------------------------------------
+class Allocate(Basic):
+    """
+    Represents memory allocation (usually of an array) for code generation.
+    This is relevant to low-level target languages, such as C or Fortran,
+    where the programmer must take care of heap memory allocation.
+
+    Parameters
+    ----------
+    variable : pyccel.ast.core.Variable
+        The typed variable (usually an array) that needs memory allocation.
+
+    shape : int or iterable or None
+        Shape of the array after allocation (None for scalars).
+
+    order : str {'C'|'F'}
+        Ordering of multi-dimensional array after allocation
+        ('C' = row-major, 'F' = column-major).
+
+    status : str {'allocated'|'unallocated'|'unknown'}
+        Variable allocation status at object creation.
+
+    Notes
+    -----
+    An object of this class is immutable, although it contains a reference to a
+    mutable Variable object.
+
+    """
+    def __new__(cls, *args, **kwargs):
+
+        return Basic.__new__(cls)
+
+    # ...
+    def __init__(self, variable, *, shape, order, status):
+
+        if not isinstance(variable, Variable):
+            raise TypeError("Can only allocate a 'Variable' object, got {} instead".format(type(variable)))
+
+        if not variable.allocatable:
+            raise ValueError("Variable must be allocatable")
+
+        if shape and not isinstance(shape, (int, tuple, list)):
+            raise TypeError("Cannot understand 'shape' parameter of type '{}'".format(type(shape)))
+
+        if variable.rank != len(shape):
+            raise ValueError("Incompatible rank in variable allocation")
+
+        # TODO [YG, 28.10.2020] Is this a good place to raise a Pyccel error?
+        #                       And if so, how do we obtain a bounding box?
```

Comment:
If we kept the fst after the syntactic step in all the classes that inherit from `ast.base.Base` then this information would be available outside of the semantic stage. Currently almost all errors are handled from the SemanticParser so that the bounding box is available, but that would require adding the if block everytime we instantiate `Allocate`.
Given that there are only 2 places in the code where we do this, maybe that's not a problem?

Reply:
I see. Then the best strategy is probably not to raise a Pyccel error from an AST node, but rather just standard Python exceptions (in order to ensure consistency of our code). The caller of this class constructor (the semantic parser) should perform a check and raise a Pyccel error if needed.

Reply:
Done in 5dc3b16